contracts/stellar-interchain-token-service/src/migrate.rs

```diff
@@ -0,0 +1,101 @@
+use stellar_axelar_std::interfaces::CustomMigratableInterface;
+use stellar_axelar_std::{contracttype, soroban_sdk, Address, BytesN, Env, String, Vec};
+use stellar_upgrader::UpgraderClient;
+
+use crate::error::ContractError;
+use crate::flow_limit::current_epoch;
+use crate::storage::TokenIdConfigValue;
+use crate::types::TokenManagerType;
+use crate::{storage, InterchainTokenService};
+
+#[contracttype]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CustomMigrationData {
+    pub upgrader_client: Address,
+    pub new_version: String,
+    pub token_ids: Vec<BytesN<32>>,
+    pub new_token_manager_wasm_hash: BytesN<32>,
+    pub new_interchain_token_wasm_hash: BytesN<32>,
+}
+
+pub mod legacy_storage {
+    use stellar_axelar_std::{contractstorage, contracttype, soroban_sdk, BytesN};
+
+    #[contractstorage]
+    enum LegacyDataKey {
+        #[temporary]
+        #[value(i128)]
+        FlowOut { flow_key: FlowKey },
+
+        #[temporary]
+        #[value(i128)]
+        FlowIn { flow_key: FlowKey },
+    }
+
+    #[contracttype]
+    #[derive(Clone, Debug)]
+    pub struct FlowKey {
+        pub token_id: BytesN<32>,
+        pub epoch: u64,
+    }
+}
+
+impl CustomMigratableInterface for InterchainTokenService {
+    type MigrationData = CustomMigrationData;
+    type Error = ContractError;
+
+    fn __migrate(env: &Env, migration_data: Self::MigrationData) -> Result<(), Self::Error> {
+        let CustomMigrationData {
+            upgrader_client,
+            new_version,
+            token_ids,
+            new_token_manager_wasm_hash,
+            new_interchain_token_wasm_hash,
+        } = migration_data;
+
+        storage::set_token_manager_wasm_hash(env, &new_token_manager_wasm_hash);
+        storage::set_interchain_token_wasm_hash(env, &new_interchain_token_wasm_hash);
+
+        let current_epoch = current_epoch(env);
+        let upgrader_client = UpgraderClient::new(env, &upgrader_client);
+
+        for token_id in token_ids.into_iter() {
+            let TokenIdConfigValue {
+                token_address: interchain_token,
+                token_manager,
+                token_manager_type,
+            } = storage::try_token_id_config(env, token_id.clone())
+                .ok_or(ContractError::InvalidTokenId)?;
+
+            upgrader_client.upgrade(
+                &token_manager,
+                &new_version,
+                &new_token_manager_wasm_hash,
+                &stellar_axelar_std::vec![env, ().into()],
```

Comment:
This interface doesn't actually make sense. `Upgradable` contracts only take one migration data arg. So the upgrader shouldn't be requiring a list of args I think, and the usage should be more like the following
```suggestion
                &().into(),
```